File: lua/glean/baseline_test.lua
-- Tier 1 tests for glean.baseline (pure reviewed-baseline algebra).
-- Run with: nvim -l lua/glean/baseline_test.lua
local this_script = debug.getinfo(1, "S").source:sub(2)
local this_dir = this_script:match("(.+)/[^/]+$") or "."
local lua_root = this_dir:match("(.+)/[^/]+$") or "."
package.path = lua_root .. "/?.lua;" .. lua_root .. "/?/init.lua;" .. package.path
local baseline = require("glean.baseline")
local h = require("glean.testutil").new()

local function lines(s)
  local out = {}
  for line in (s .. "\n"):gmatch("([^\n]*)\n") do
    out[#out + 1] = line
  end
  return out
end

-- Render seen-ness of the displayed diff D = diff(base, worktree) as a compact
-- string: one entry per displayed add/del line, e.g. "+3:seen -2:unseen".
local function render(base, reviewed, worktree)
  local sets = baseline.seen_sets(base, reviewed, worktree)
  local parts = {}
  for _, op in ipairs(baseline.align(base, worktree)) do
    if op.kind == "add" then
      parts[#parts + 1] = ("+%d:%s"):format(op.b_lnum, sets.add[op.b_lnum] and "seen" or "unseen")
    elseif op.kind == "del" then
      parts[#parts + 1] = ("-%d:%s"):format(op.a_lnum, sets.del[op.a_lnum] and "seen" or "unseen")
    end
  end
  return table.concat(parts, " ")
end

-- All displayed adds/dels of diff(base, worktree), i.e. "mark everything".
local function all_sel(base, worktree)
  local sel = { add = {}, del = {} }
  for _, op in ipairs(baseline.align(base, worktree)) do
    if op.kind == "add" then
      sel.add[#sel.add + 1] = op.b_lnum
    elseif op.kind == "del" then
      sel.del[#sel.del + 1] = op.a_lnum
    end
  end
  return sel
end

local function eq_lines(name, actual, expected)
  h.assert_eq(name, table.concat(actual, "|"), table.concat(expected, "|"))
end

-- Baseline scenario: a three-line insertion into a five-line file, marked seen,
-- then one of the three edited. Only the edited line comes back unseen.
local base = lines("a\nb\nc\nd\ne")
local function edited_case(name, edit_idx, new_text)
  local wt = lines("a\nb\nx\ny\nz\nc\nd\ne")
  local reviewed = baseline.mark(base, base, wt, all_sel(base, wt))
  eq_lines("mark " .. name .. ": R == W", reviewed, wt)
  h.assert_eq("marked " .. name .. ": all seen", render(base, reviewed, wt), "+3:seen +4:seen +5:seen")
  local wt2 = vim.deepcopy(wt)
  wt2[edit_idx] = new_text
  h.assert_eq(
    "edited " .. name,
    render(base, reviewed, wt2),
    ("+3:%s +4:%s +5:%s"):format(
      edit_idx == 3 and "unseen" or "seen",
      edit_idx == 4 and "unseen" or "seen",
      edit_idx == 5 and "unseen" or "seen"
    )
  )
  -- Reverting the edit restores seen-ness: the line matches R again.
  h.assert_eq("reverted " .. name, render(base, reviewed, wt), "+3:seen +4:seen +5:seen")
end
edited_case("first line of run", 3, "X")
edited_case("middle of run", 4, "Y")
edited_case("last line of run", 5, "Z")

-- Inserting a line into the middle of a marked run: only the new line is
-- unseen; both halves stay seen. Block records cannot express this.
do
  local wt = lines("a\nb\nx\ny\nz\nc\nd\ne")
  local reviewed = baseline.mark(base, base, wt, all_sel(base, wt))
  local wt2 = lines("a\nb\nx\nNEW\ny\nz\nc\nd\ne")
  h.assert_eq("insert into marked run", render(base, reviewed, wt2), "+3:seen +4:unseen +5:seen +6:seen")
end

-- Shifting a marked region (text inserted above it): it stays seen, because
-- the diff follows it rather than matching it by position. A wholesale *move*
-- across the file is not covered: vim.diff is not move-aware, so the region
-- reads as a fresh insertion elsewhere and is unseen.
do
  local wt = lines("a\nb\nx\ny\nc\nd\ne")
  local reviewed = baseline.mark(base, base, wt, all_sel(base, wt))
  local wt2 = lines("a\nNEW\nb\nx\ny\nc\nd\ne")
  h.assert_eq("shifted marked region", render(base, reviewed, wt2), "+2:unseen +4:seen +5:seen")
end

-- Two identical new lines far apart: marking one leaves the other unseen.
do
  local wt = lines("a\ndup\nb\nc\nd\ndup\ne")
  local reviewed = baseline.mark(base, base, wt, { add = { 2 } })
  h.assert_eq("duplicate lines", render(base, reviewed, wt), "+2:seen +6:unseen")
end

-- Deletions: marking a del removes the line from R and it reads seen.
do
  local wt = lines("a\nb\nd\ne")
  h.assert_eq("unmarked del", render(base, base, wt), "-3:unseen")
  local reviewed = baseline.mark(base, base, wt, { del = { 3 } })
  eq_lines("mark del: R == W", reviewed, wt)
  h.assert_eq("marked del", render(base, reviewed, wt), "-3:seen")
  -- Restoring the deleted line in the work tree: D is empty, and R (which
  -- lacks the line) is simply not consulted about it.
  h.assert_eq("del restored in W", render(base, reviewed, base), "")
end

-- A del and an add in one selection, marked together.
do
  local wt = lines("a\nb\nC\nd\ne")
  local reviewed = baseline.mark(base, base, wt, all_sel(base, wt))
  h.assert_eq("changed line", render(base, reviewed, wt), "-3:seen +3:seen")
  eq_lines("changed line: R == W", reviewed, wt)
end

-- Partial selection: mark only one of two independent changes.
do
  local wt = lines("a\nP\nb\nc\nd\ne\nQ")
  local reviewed = baseline.mark(base, base, wt, { add = { 7 } })
  h.assert_eq("partial mark", render(base, reviewed, wt), "+2:unseen +7:seen")
end

-- mark is idempotent; unmark inverts mark exactly.
do
  local wt = lines("a\nP\nc\nd\ne\nQ") -- adds P, drops b, appends Q
  local sel = all_sel(base, wt)
  local r1 = baseline.mark(base, base, wt, sel)
  local r2 = baseline.mark(base, r1, wt, sel)
  eq_lines("mark idempotent", r2, r1)
  eq_lines("unmark inverts mark", baseline.unmark(base, r1, wt, sel), base)
  h.assert_eq("unmarked: all unseen", render(base, baseline.unmark(base, r1, wt, sel), wt), render(base, base, wt))

  -- Same for a partial selection.
  local partial = { add = { 2 }, del = { 2 } }
  local p1 = baseline.mark(base, base, wt, partial)
  local p2 = baseline.mark(base, p1, wt, partial)
  eq_lines("partial mark idempotent", p2, p1)
  eq_lines("partial unmark inverts", baseline.unmark(base, p1, wt, partial), base)
end

-- Empty files at either endpoint.
do
  local wt = lines("a\nb")
  local reviewed = baseline.mark({}, {}, wt, all_sel({}, wt))
  eq_lines("mark from empty base", reviewed, wt)
  h.assert_eq("new file all seen", render({}, reviewed, wt), "+1:seen +2:seen")
  local gone = baseline.mark(base, base, {}, all_sel(base, {}))
  eq_lines("mark whole-file deletion", gone, {})
end

h.finish()

File: node/core/baseline.test.ts
import { describe, expect, it } from "vitest";
import { align, markAdds, unmarkAdds, unseenAdds } from "./baseline.ts";
import type { WorktreeLnum } from "./types.ts";

const lines = (s: string) => s.split("\n");
const wl = (...n: number[]) => n as WorktreeLnum[];

function render(base: string[], reviewed: string[], wt: string[]): string {
  const unseen = unseenAdds(reviewed, wt);
  return align(base, wt)
    .flatMap((op) =>
      op.kind === "add"
        ? [
            `+${op.bLnum}:${unseen.has(op.bLnum as WorktreeLnum) ? "unseen" : "seen"}`,
          ]
        : [],
    )
    .join(" ");
}

function allAdds(base: string[], wt: string[]): WorktreeLnum[] {
  return align(base, wt).flatMap((op) =>
    op.kind === "add" ? [op.bLnum as WorktreeLnum] : [],
  );
}

function expectAddOnly(head: string[], reviewed: string[]) {
  expect(align(head, reviewed).filter((op) => op.kind === "del")).toEqual([]);
}

const base = lines("a\nb\nc\nd\ne");

describe("baseline", () => {
  it.each([
    ["first line of run", 3, "X"],
    ["middle of run", 4, "Y"],
    ["last line of run", 5, "Z"],
  ])("editing the %s makes only it unseen", (_, idx, text) => {
    const wt = lines("a\nb\nx\ny\nz\nc\nd\ne");
    const reviewed = markAdds(base, wt, allAdds(base, wt));
    expect(reviewed).toEqual(wt);
    expect(render(base, reviewed, wt)).toBe("+3:seen +4:seen +5:seen");
    const wt2 = [...wt];
    wt2[idx - 1] = text;
    const s = (k: number) => (k === idx ? "unseen" : "seen");
    expect(render(base, reviewed, wt2)).toBe(
      `+3:${s(3)} +4:${s(4)} +5:${s(5)}`,
    );
    expect(render(base, reviewed, wt)).toBe("+3:seen +4:seen +5:seen");
  });

  it("insert into a marked run", () => {
    const wt = lines("a\nb\nx\ny\nz\nc\nd\ne");
    const reviewed = markAdds(base, wt, allAdds(base, wt));
    expect(render(base, reviewed, lines("a\nb\nx\nNEW\ny\nz\nc\nd\ne"))).toBe(
      "+3:seen +4:unseen +5:seen +6:seen",
    );
  });

  it("a shifted marked region stays seen", () => {
    const wt = lines("a\nb\nx\ny\nc\nd\ne");
    const reviewed = markAdds(base, wt, allAdds(base, wt));
    expect(render(base, reviewed, lines("a\nNEW\nb\nx\ny\nc\nd\ne"))).toBe(
      "+2:unseen +4:seen +5:seen",
    );
  });

  it("duplicate lines far apart", () => {
    const wt = lines("a\ndup\nb\nc\nd\ndup\ne");
    expect(render(base, markAdds(base, wt, wl(2)), wt)).toBe(
      "+2:seen +6:unseen",
    );
  });

  it("deletions leave R untouched", () => {
    const wt = lines("a\nb\nd\ne");
    const reviewed = markAdds(base, wt, allAdds(base, wt));
    expect(reviewed).toEqual(base);
    expectAddOnly(base, reviewed);
  });

  it("duplicate-heavy fixture never shrinks R", () => {
    const head = lines(
      "$$;\n\n--\n-- Name: A\n--\n\nBODY A\n\n--\n-- Name: B\n--\n\nBODY B",
    );
    const wt = lines("$$;\n\n--\n-- Name: B\n--\n\nBODY B\nNEW");
    const reviewed = markAdds(head, wt, allAdds(head, wt));
    expect(reviewed.length).toBeGreaterThanOrEqual(head.length);
    expectAddOnly(head, reviewed);
    expect(render(head, reviewed, wt)).toBe("+8:seen");
  });

  it("partial selection", () => {
    const wt = lines("a\nP\nb\nc\nd\ne\nQ");
    expect(render(base, markAdds(base, wt, wl(7)), wt)).toBe(
      "+2:unseen +7:seen",
    );
  });

  it("mark is idempotent and unmark inverts it", () => {
    const wt = lines("a\nP\nc\nd\ne\nQ");
    const sel = allAdds(base, wt);
    const r1 = markAdds(base, wt, sel);
    expect(markAdds(r1, wt, sel)).toEqual(r1);
    expectAddOnly(base, r1);
    const undone = unmarkAdds(base, r1, wt, sel);
    expect(undone).toEqual(base);
    expect(render(base, undone, wt)).toBe(render(base, base, wt));
    const p1 = markAdds(base, wt, wl(2));
    expect(markAdds(p1, wt, wl(2))).toEqual(p1);
    expect(unmarkAdds(base, p1, wt, wl(2))).toEqual(base);
  });

  it("empty files at either endpoint", () => {
    const wt = lines("a\nb");
    const reviewed = markAdds([], wt, allAdds([], wt));
    expect(reviewed).toEqual(wt);
    expect(render([], reviewed, wt)).toBe("+1:seen +2:seen");
    expect(markAdds(base, [], allAdds(base, []))).toEqual(base);
  });
});

describe("align (Myers)", () => {
  function lcs(a: string[], b: string[]): number {
    const dp = Array.from({ length: a.length + 1 }, () =>
      new Array<number>(b.length + 1).fill(0),
    );
    for (let i = 1; i <= a.length; i++)
      for (let j = 1; j <= b.length; j++)
        dp[i]![j] =
          a[i - 1] === b[j - 1]
            ? dp[i - 1]![j - 1]! + 1
            : Math.max(dp[i - 1]![j]!, dp[i]![j - 1]!);
    return dp[a.length]![b.length]!;
  }

  it("reconstructs both sides with a minimal script", () => {
    let seed = 7;
    const rnd = (k: number) => {
      seed = (seed * 1103515245 + 12345) % 2 ** 31;
      return seed % k;
    };
    for (let t = 0; t < 300; t++) {
      const a = Array.from({ length: rnd(12) }, () => "abc"[rnd(3)]!);
      const b = Array.from({ length: rnd(12) }, () => "abc"[rnd(3)]!);
      const ops = align(a, b);
      expect(ops.flatMap((o) => (o.kind === "add" ? [] : [o.text]))).toEqual(a);
      expect(ops.flatMap((o) => (o.kind === "del" ? [] : [o.text]))).toEqual(b);
      expect(ops.filter((o) => o.kind === "context").length).toBe(lcs(a, b));
    }
  });
});

describe("linediff cap", () => {
  it("falls back to del-then-add for wholesale rewrites quickly", async () => {
    const { alignLines } = await import("./linediff.ts");
    const a = Array.from({ length: 20000 }, (_, i) => `a${i}`);
    const b = Array.from({ length: 20000 }, (_, i) => `b${i}`);
    const t = Date.now();
    const ops = alignLines(a, b);
    expect(Date.now() - t).toBeLessThan(1000);
    expect(ops.filter((o) => o.kind === "del").length).toBe(20000);
    expect(ops[20000]?.kind).toBe("add");
  });
});
